ruby/phase-1.md

```diff
@@ -0,0 +1,598 @@
+end
+
+class MyClass
+  def my_method
+    puts 'Hello Monkeys!'
+  end
+
+  def new_method
+    puts 'Hello New World!'
+  end
+end
+
+my = MyClass.new
+my.my_method
+# => Hello Monkeys!
+my.new_method
+# => Hello New World
+```
+
+### Concurrency
+
+Ruby can be a tricky language for running concurrent applications. It supports kernel threads,
+however it has a Global Interpreter Lock (GIL) in place which prevents any two threads to be
+executed at the same time (making it effectively single-threaded). The good news is that GIL does
+allow IO-bound instructions to be executed in parallel.
+
+While threads can offer a great performance boost in applications with a significant IO reliance. In
+order to effectively utilise a multi-core processor a Ruby application has to be executed using
+multiple processes.
+
+#### Fibers
+
+Ruby also has Fibers that allow light weight cooperative concurrency. These are code blocks that can
+be paused (from inside) and resumed (from outside). These are really useful for controlling the
+execution of workflow code. However unlike threads Fibers are never preempted and their scheduling
+has to be done manually by the code:
+
+```ruby
+fiber = Fiber.new do
+  i = 0
+  loop do
+    i += 1
+    puts i
+    Fiber.yield
+  end
+end
+
+fiber.resume
+# => 1
+fiber.resume
+# => 2
+fiber.resume
+# => 3
+```
+
+*NOTE: Each `fiber.resume` call blocks the main thread until a Fiber `yield`s back control.*
+
+#### Reactors
+
+Somewhat less popular, but still very viable are libraries that implement a reactor pattern for
+building concurrent applications (e.g. Async or EventMachine). These libraries allow users to
+concurrently execute IO operations.
+
+There are however quite a few big limitations to this approach:
+
+- You need to use very specific libraries to perform IO calls this way (there isn't one for gRPC)
+- Any CPU-bound instructions block the reactor and therefore need to be executed in small bursts per
+  reactor tick
+- Any user-defined code can block the whole thing and has to be wrapped in threads anyways
+
+
+## A basic end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+class MyActivity < Temporal::Activity
+  def execute(name)
+    puts "Hello World, #{name}!"
+  end
+end
+
+class MyWorkflow < Temporal::Workflow
+  def execute(name)
+    # This part is outside of the scope of the 1st phase of the proposal
+    MyActivity.execute!(name, activity: { task_queue: 'my-queue' })
+  end
+end
+
+config = Temporal::Configuration.new(
+  url: 'localhost:7233'
+)
+
+client = Temporal::Client.new(config)
+client.start_workflow(MyWorkflow, 'Alice', workflow: {
```

Comment:
I'm bothered by the 3x repetition of "workflow". Can the last one be removed or replaced with `options:`?

Reply:
`MyWorkflow` is user defined and doesn't have to have `Workflow` in it. As for the `workflow` — I had `options:` originally, however switched to `workflow:` to minimise collision with user-defined keyword arguments.

We can potentially rename `#start_workflow` to just `#start` since it shouldn't clash with other available operations

Reply:
I think we want to keep `start_workflow` because all top-level client functions need to include what they apply to in the name (e.g. async activity calls will have the word `activity` in them as will `start_activity` if/when we ever get top-level activities).

As for the option name, I do think `options` is clearer and it is unfortunate that the not-yet-implemented only-in-Ruby potential future approach of supporting kwargs has now adversely affected this param name that will probably be the most commonly used named parameter in the entire SDK.

Having said that, I understand the kwarg name clash concern. Granted one might expect if it's so easy for them to provide a key-value set of options as a single arg to our call, they can surely provide a key-value set of fields as a single arg to their call.

Reply:
My inclination would be to prioritize clearer `options:` name (something that's always used) over avoiding collision (something that would only be relevant sometimes, and has an easy workaround).

Reply:
I don't know if I like it better, but what about a handle approach like:
```ruby
client.workflow(MyWorkflow, options: {id: 'foo'}).start('Alice')
```

You could still even support kwargs if you'd like, but it's clearer they are for arguments. You could have similar constructs for attaching to an already running workflow:
```ruby
client.workflow(MyWorkflow, options: {id: 'foo', run_id: '...'}).result()
client.workflow(MyWorkflow, options: {id: 'foo', run_id: '...'}).describe()
```

Reply:
This is close to what we do in Java with "stubs". Of course, you might not want to mix unstarted with expected to be running because that can confuse things. But this works:

```ruby
stub = client.workflow_stub(MyWorkflow)
handle = stub.start('Alice')

untyped_handle = client.workflow_handle(options: { id: 'foo' })

handle = client.workflow_handle_for(MyWorkflow, options { id: 'foo' })
```

If you wanted to go all the way w/ typing, you instead may have:

```ruby
stub = client.workflow_stub(MyWorkflow)
# Start + result()
result = stub.execute('Alice')

handle = client.start_workflow(MyWorkflow.instance_method(:execute), 'Alice')
```

Assuming the type checker works, but I can understand if that's ugly. There are plenty of other approaches too.

Reply:
In TS we just put the args in the options parameter:
```ts
client.start(someWorkflow, { workflowId: '...', taskQueue: 'q', args: ['foo', 'bar'] })
```

But TS type system is strong enough to infer the type of `args` from the `someWorkflow` signature.